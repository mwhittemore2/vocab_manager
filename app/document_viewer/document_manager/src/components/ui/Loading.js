import React from "react"

/**
 * Renders a loading message for any component that
 * is experiencing high network latency.
 */
export class Loading extends React.Component {
    render(){
        return (
            <b>Loading...</b>
        )
    }
}