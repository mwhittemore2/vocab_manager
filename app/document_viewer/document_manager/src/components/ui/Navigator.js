import PropTypes from "prop-types"
import React from "react"
import C from '../../constants'

export const Navigator = ({cursor, jumpToPage, results, viewer}) =>
    <div id={viewer.concat("-navigator")}>
        <button id={viewer.concat("-previous")} onClick={cursor(C.PREVIOUS_PAGE, results.currPage)}>
            Previous
        </button>
        <div id={viewer.concat("-current")}>
           <input onKeyDown={(e) => jumpToPage(e)}
                  type="text"
                  placeholder={results.currPage}/>
        </div>
        <button id={viewer.concat("-next")} onClick={cursor(C.NEXT_PAGE, results.currPage)}>
            Next
        </button>
    </div>

Navigator.propTypes = {
    cursor: PropTypes.func,
    jumpToPage: PropTypes.func,
    results: PropTypes.object,
    viewer: PropTypes.string
}