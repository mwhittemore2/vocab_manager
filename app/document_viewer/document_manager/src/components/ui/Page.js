import React from "react"
import PropTypes from 'prop-types'
import { Word } from './Word'

const buildLineKey = (index, page) => {
    let domain = "line"
    let separator = "-"
    let key = domain + separator + page.currPage.toString() + separator + index.toString()
    return key
}

const buildWordKey = (index, word) => {
    let domain = "word"
    let separator = "-"
    let key = domain + separator + index.toString() + separator + word
    return key
}

const buildWord = (index, line, page, word) => {
    return {line: line, page: page.currPage, pos: index, text: word}
}

const canColor = (line, pos, page) => {
    let separator = ":"
    let word = page.currPage.toString() + separator + line.toString() + separator + pos.toString()
    let decision = page.selected.has(word)
    return decision
}

export const Line = ({line, num, page, selectWord}) => {
    return (
        <div>
            {line.map((word, index) => 
                <Word canColor={canColor(num, index, page)}
                      key={buildWordKey(index, word)}
                      selectWord={selectWord}
                      word={buildWord(index, num, page, word)}/>)}
            <br></br>
        </div>       
    )    
}

export const Page = ({lines={}, selectWord=f=>f}) =>
    <div className="page">
        {lines.words.map((line, index) =>
            <Line key ={buildLineKey(index, lines)}
                  line={line}
                  num={index}
                  page={lines}
                  selectWord={selectWord}/>)}
    </div>

Line.propTypes = {
    line: PropTypes.array,
    num: PropTypes.number,
    page: PropTypes.object,
    selectWord: PropTypes.func
}

Page.propTypes = {
    lines: PropTypes.object,
    selectWord: PropTypes.func
}