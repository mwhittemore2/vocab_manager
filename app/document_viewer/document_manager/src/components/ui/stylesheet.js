const stylesheet = {
    buttonDivider: "button-divider",
    chooseOption: "choose-option",
    navigator: "navigator",
    option: "option",
    optionHeader: "option-header"
}

export default stylesheet