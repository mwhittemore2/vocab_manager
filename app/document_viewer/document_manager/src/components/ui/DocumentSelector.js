import PropTypes from 'prop-types'
import React from "react"
import C from '../../constants'
import { Loading } from './Loading'

const buildKey = (doc, index) => {
    let domain = "docChoice"
    let separator = "-"
    let key = domain + separator + index.toString() + separator + doc
    return key
}

export const DocumentChoice = ({doc, selector}) =>
    <div onClick={() => selector(doc)}>
        {doc.title}
    </div>

export const DocumentIterator = ({documents, selectDocument}) =>
    <div id="document-selector">
        {documents.sort((a, b) => (a.title > b.title) ? 1 : -1).map((doc, index) =>
            <DocumentChoice key={buildKey(doc.title, index)}
                            doc={doc}
                            selector={selectDocument}/>
        )}
    </div>

 export class DocumentSelector extends React.Component {
    render(){
        let documents = this.props.documents
        let getDocuments = this.props.getDocuments
        let interaction = this.props.interaction
        let loaded = this.props.loaded[C.DOCUMENT_SELECTOR]
        let selectDocument = this.props.selectDocument
        if (interaction === C.DOCUMENT_SELECTOR){
            if (loaded){
                return (
                    <DocumentIterator documents={documents}
                                      selectDocument={selectDocument}/>
                )
            }
            else{
                if (documents.length <= 0){
                    getDocuments()
                }
                return (
                    <Loading />
                )
            }
        }
        return (
            <div id={"document-selector"}/>
        )
    }
}

DocumentChoice.propTypes = {
    doc: PropTypes.object,
    selector: PropTypes.func
}

DocumentIterator.propTypes = {
    documents: PropTypes.array,
    selectDocument: PropTypes.func
}

DocumentSelector.propTypes = {
    documents: PropTypes.array,
    getDocuments: PropTypes.func,
    interaction: PropTypes.string,
    loaded: PropTypes.object,
    selectDocument: PropTypes.func
}

DocumentSelector.defaultProps = {
    documents: [],
    getDocuments: f=>f,
    interaction: C.DOCUMENT_SELECTOR,
    loaded: {},
    selectDocument: f=>f
}

export default DocumentSelector