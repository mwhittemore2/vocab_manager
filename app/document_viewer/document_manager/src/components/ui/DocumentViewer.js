import PropTypes from "prop-types"
import React from "react"
import { Navigator } from './Navigator'
import { Page } from './Page'
import stylesheet from './stylesheets/common'
import C from '../../constants'

const displayID = "document-viewer"

/**
 * Manages the current page that the user is reading.
 * 
 * @param {func} cursor Function for moving to previous or next page.
 * @param {func} jumpToPage Function for moving to a specified page.
 * @param {object} lines The text of the currently displayed page.
 * @param {func} selectWord Adds the user-selected word to the translation 
 *                          queue.
 * @return {html} The HTML representation of the current page.
 */
export const Display = ({cursor, jumpToPage, lines, selectWord}) =>
    <div id={displayID} className={stylesheet.displayDocumentViewer}>
        <Page lines={lines} selectWord={selectWord} />
        <Navigator cursor={cursor}
                   jumpToPage={jumpToPage}
                   results={lines}
                   viewer={displayID}/>
    </div>

/**
 * Renders the interface for reading and traversing a document.
 */
export class DocumentViewer extends React.Component{
    render(){
        let interaction = this.props.interaction
        if(interaction === C.DOCUMENT_VIEWER){
            let cursor = this.props.cursor
            let jumpToPage = this.props.jumpToPage
            let lines = this.props.lines
            let selectWord = this.props.selectWord
            return(
                <Display cursor={cursor}
                         jumpToPage={jumpToPage}
                         lines={lines}
                         selectWord={selectWord}/>
            )
        }
        return(
            <div id={displayID}></div>
        )
    }
}

Display.propTypes = {
    cursor: PropTypes.func,
    jumpToPage: PropTypes.func,
    lines: PropTypes.object,
    selectWord: PropTypes.func
}

DocumentViewer.propTypes = {
    cursor: PropTypes.func,
    interaction: PropTypes.string,
    jumpToPage: PropTypes.func,
    lines: PropTypes.object,
    selectWord: PropTypes.func
}

DocumentViewer.defaultProps = {
    cursor: f=>f,
    interaction: C.DOCUMENT_SELECTOR,
    jumpToPage: f=>f,
    lines: {},
    selectWord: f=>f
}

export default DocumentViewer