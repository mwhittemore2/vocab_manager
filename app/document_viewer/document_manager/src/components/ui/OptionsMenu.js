import PropTypes from "prop-types"
import React from "react"
import stylesheet from './stylesheets/common'
import C from "../../constants"

const optionsMenuID = "document-viewer-options-menu"

export const DisplayOptions = ({selectOption}) =>
    <div id={optionsMenuID} className={stylesheet.option}>
        <b className={stylesheet.optionHeader}>Document Viewer Options</b>
        <br></br>
        <br></br>
        <ul className={stylesheet.chooseOption}>
            <div onClick={() => selectOption(C.DOC_VIEWER_OPTIONS.TRANSLATION_COORDINATOR)}>
                <li>Select words to translate</li>
            </div>
            <br></br>
            <div onClick={() => window.alert("This feature isn't implemented yet")}>
                <li>Manage bookmarks</li>
            </div>
        </ul>
    </div>

export class OptionsMenu extends React.Component{
    render(){
        let interaction = this.props.interaction
        if(interaction === C.DOCUMENT_VIEWER){
            let option = this.props.option
            if(option === C.DOC_VIEWER_OPTIONS.DEFAULT){
                let selectOption = this.props.selectOption
                return(
                    <DisplayOptions selectOption={selectOption}/>
                )
            }
        }
        return(
            <div id={optionsMenuID}></div>
        )
    }
}

DisplayOptions.propTypes = {
    selectOption: PropTypes.func
}

OptionsMenu.propTypes = {
    interaction: PropTypes.string,
    option: PropTypes.string,
    selectOption: PropTypes.func
}

/*
OptionsMenu.defaultProps = {
    interaction: C.DOCUMENT_VIEWER,
    option: C.DOC_VIEWER_OPTIONS.DEFAULT,
    selectOption: f=>f
}*/

export default OptionsMenu