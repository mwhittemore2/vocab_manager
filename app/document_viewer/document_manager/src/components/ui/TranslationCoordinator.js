import PropTypes from "prop-types"
import React from "react"
import C from "../../constants"
import stylesheet from './stylesheet'
import { grammar } from "../../data/grammar"

const coordinatorID = "translation-coordinator"

const customTextID = "translation-coordinator-custom-text-manager"

const customTextSubmit = "translation-coordinator-custom-text-submit"

const endpointManagerID = "translation-coordinator-endpoint-manager"

const queueManagerID = "translation-coordinator-queue-manager"

const textManagerID = "translation-coordinator-text-manager"

const buildTextManagerKey = (index, text) => {
    let domain = "text-manager"
    let separator = "-"
    let key = domain + separator + index.toString() + separator + text
    return key
}

const processTranslationRequest = (currState, translate) => {
    if(currState.searchPhrase.length > 0){
        translate()
    }
    else{
        let errorMessage = "Translation Queue must have content to be translated."
        window.alert(errorMessage)
    }
}

const submitCustomText = (e, addText) => {
    let enterKey = C.KEYBOARD_INPUT.ENTER
    if(e.keyCode === enterKey){
        let customText = e.target.value
        if(customText){
            addText(customText)
            document.getElementById(customTextSubmit).value=''
        }
        else{
            let errorMessage = "Enter text to add to the translation queue"
            window.alert(errorMessage)
        }
    }
}

export const CustomText = ({addText}) =>
    <div id={customTextID}>
        <input id={customTextSubmit}
               onKeyDown={(e) => submitCustomText(e, addText)}
               type="text"
               placeholder="Enter text here"/>
    </div>

export const EndpointManager = ({setEndpoint}) => 
    <div id={endpointManagerID}>
        Set text boundaries
        <br></br>
        <button className={stylesheet.buttonDivider}
                onClick={() => setEndpoint(C.TEXT_START)}>Start</button>
        <button onClick={() => setEndpoint(C.TEXT_FINISH)}>End</button>
        <br></br>
        <br></br>
    </div>

export const QueueManager = ({clearQueue}) => 
    <div id={queueManagerID}>
        Manage the translation queue
        <br></br>
        <button onClick={() => clearQueue()}>Clear Queue</button>
        <br></br>
        <br></br>
    </div>

export const Punctuation = ({addText, punct}) => 
    <button className={stylesheet.buttonDivider}
            onClick={() => addText(punct)}>{punct}</button>

export const Spacing = ({addText, description, spacing}) =>
    <button onClick={() => addText(spacing)}>{description}</button>

export const TextManager = ({addText}) => 
    <div id={textManagerID}>
        Punctuation
        <br></br>
        {grammar.default.punctuation.map((punct, index) => 
            <Punctuation addText={addText}
                         key={buildTextManagerKey(index, punct)}
                         punct={punct}/>)}
        <br></br>
        <br></br>
        Spacing
        <br></br>
        {Object.keys(grammar.default.spacing).sort().map((description, index) =>
            <Spacing addText={addText}
                     description={description}
                     key={buildTextManagerKey(index, description)}
                     spacing={grammar.default.spacing[description]}/>)}
        <br></br>
        <br></br>
        Custom Text
        <CustomText addText={addText}/>
    </div>

export class TranslationCoordinator extends React.Component{
    render(){
        let interaction = this.props.interaction
        if(interaction === C.DOCUMENT_VIEWER){
            let option = this.props.option
            if(option === C.DOC_VIEWER_OPTIONS.TRANSLATION_COORDINATOR){
                let addText = this.props.addText
                let clearQueue = this.props.clearQueue
                let setEndpoint = this.props.setEndpoint
                let translate = this.props.translate
                let translations = this.props.translations
                return(
                    <div id={coordinatorID} className={stylesheet.option}>
                        <b className={stylesheet.optionHeader}>Translation Coordinator</b>
                        <br></br>
                        <br></br>
                        <EndpointManager setEndpoint={setEndpoint}/>
                        <QueueManager clearQueue={clearQueue}/>
                        <TextManager addText={addText}/>
                        <br></br>
                        <br></br>
                        <button onClick={() => {
                            processTranslationRequest(translations, translate)
                        }}>Get Translations</button>
                    </div>
                )
            }
        }
        return(
            <div id={coordinatorID}></div>
        )
    }
}

CustomText.propTypes = {
    addText: PropTypes.func
}

EndpointManager.propTypes = {
    setEndpoint: PropTypes.func
}

QueueManager.propTypes = {
    clearQueue: PropTypes.func
}

Punctuation.propTypes = {
    addText: PropTypes.func,
    punct: PropTypes.string
}

Spacing.propTypes = {
    addText: PropTypes.func,
    description: PropTypes.string,
    spacing: PropTypes.string
}

TextManager.propTypes = {
    addText: PropTypes.func
}

TranslationCoordinator.propTypes = {
    addText: PropTypes.func,
    clearQueue: PropTypes.func,
    interaction: PropTypes.string,
    option: PropTypes.string,
    setEndpoint: PropTypes.func,
    translate: PropTypes.func,
    translations: PropTypes.object
}

export default TranslationCoordinator