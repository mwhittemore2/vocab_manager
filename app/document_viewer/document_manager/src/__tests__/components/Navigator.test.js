import React from "react"
import { shallow } from "enzyme"
import toJSON from 'enzyme-to-json'
import { Navigator } from '../../components/ui/Navigator'
import C from '../../constants'

describe("Navigator", () => {
    let cursor
    let jumpToPage
    let results
    let viewer
    beforeEach(() => {
        cursor = jest.fn()
        jumpToPage = jest.fn()
        results = {
            currPage: 1
        }
        viewer = C.DOCUMENT_VIEWER
    })

    afterAll(() => {
        jest.resetAllMocks()
    })

    it("correct display", () => {
        let txt = toJSON(
                      shallow(<Navigator cursor={cursor}
                                         jumpToPage={jumpToPage}
                                         results={results}
                                         viewer={viewer}/>
                      )
                  )
        expect(txt).toMatchSnapshot()
    })

    it("previous page", () => {
        let previous = "#" + C.DOCUMENT_VIEWER + "-previous"
        shallow(<Navigator cursor={cursor}
                           jumpToPage={jumpToPage}
                           results={results}
                           viewer={viewer}/>
        )
        .find(previous)
        .simulate('click')
        expect(cursor).toBeCalled()    
    })

    it("next page", () => {
        let next = "#" + C.DOCUMENT_VIEWER + "-next"
        shallow(<Navigator cursor={cursor}
                           jumpToPage={jumpToPage}
                           results={results}
                           viewer={viewer}/>
        )
        .find(next)
        .simulate('click')
        expect(cursor).toBeCalled()
    })
})