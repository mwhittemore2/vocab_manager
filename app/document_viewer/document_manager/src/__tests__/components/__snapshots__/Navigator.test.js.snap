// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigator correct display 1`] = `
<div
  className="navigator"
  id="DOCUMENT_VIEWER-navigator"
>
  <button
    id="DOCUMENT_VIEWER-previous"
    onClick={[Function]}
  >
    Previous
  </button>
  <input
    id="DOCUMENT_VIEWER-current"
    onKeyDown={[Function]}
    placeholder={1}
    type="text"
  />
  <button
    id="DOCUMENT_VIEWER-next"
    onClick={[Function]}
  >
    Next
  </button>
</div>
`;
